py_zipkin/zipkin.py

```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
         'X-B3-Flags': '0',
         'X-B3-Sampled': '1' if zipkin_attrs.is_sampled else '0',
     }
+
+
+def is_valid_trace_id(candidate):
+    return candidate and re.match(r'(?:[0-9a-f]{16}|[0-9a-f]{32})$', candidate)
+
+
+def is_valid_span_id(candidate):
+    return candidate and re.match(r'[0-9a-f]{16}$', candidate)
+
+
+def _extract_zipkin_attrs_from_single_header(b3_header, sample_rate,
+                                             use_128bit_trace_id):
+    # b3={TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+    # where the last two fields are optional.
+    bits = b3_header.split('-')
+    if len(bits) == 1:
+        # See if it's a valid encoding of (only) sample decision
+        if bits[0] == 'false' or bits[0] == '0':
+            sample_rate = 0.0
+        elif bits[0] == 'true' or bits[0] == '1':
+            sample_rate = 100.0
+        else:
+            # Nonsense gets a None
+            return None
+        return create_attrs_for_span(
+            sample_rate=sample_rate,
+            use_128bit_trace_id=use_128bit_trace_id)
+    if not is_valid_trace_id(bits[0]) or not is_valid_span_id(bits[1]):
+        return None
+    if len(bits) == 4:
+        if not is_valid_span_id(bits[3]):
+            return None
+        parent_span_id = bits[3]
+    else:
+        parent_span_id = None
+    if len(bits) >= 3 and bits[2]:
+        # We have 1==Accept, 0==Deny, d==Debug
+        if bits[2] == 'd' or bits[2] == '1' or bits[2] == 'true':
+            is_sampled = True
+        elif bits[2] == '0' or bits[2] == 'false':
+            is_sampled = False
+        else:
+            # Not valid according to spec
+            return None
+    else:
+        # sample flag missing; means "Defer" and we're responsible for
+        # rolling fresh dice
+        is_sampled = choose_sampling(sample_rate)
+    return ZipkinAttrs(bits[0], bits[1], parent_span_id,
+                       '0',  # Flags aren't a thing in single-header
+                       is_sampled)
+
+
+def extract_zipkin_attrs_from_headers(headers, sample_rate=100.0,
+                                      use_128bit_trace_id=False):
+    """
+    Implements extraction of B3 headers per:
+        https://github.com/openzipkin/b3-propagation
+
+    The input headers can be any dict-like container that supports "in"
+    membership test and a .get() method that accepts a default value.
+
+    Returns a ZipkinAttrs instance or None
+    """
+    if 'b3' in headers:
+        return _extract_zipkin_attrs_from_single_header(
+            headers['b3'], sample_rate, use_128bit_trace_id)
+
+    # No single header, let's get old-school
```

Comment:
since you have an helper function for `_extract_zipkin_attrs_from_single_header` I'd add one also for the regular headers

Reply:
I did, but it's literally just repeating the function signature of the enclosing function.

```
    return _extract_zipkin_attrs_from_multiple_headers(
        headers, sample_rate, use_128bit_trace_id)
```

I could extract the values from headers and then call a helper function, but then it'd need to take 5 args.  Not sure that helps readability, either.